strategies/taskRankCpuUtilStrategy.go

```diff
@@ -0,0 +1,212 @@
+				colocatedTasksCpuUsageSeconds = make(map[taskIDValueT]*entities.Task)
+				allTasksCpuUsageSeconds[hostValueT(hostname)] = colocatedTasksCpuUsageSeconds
+			}
+
+			var taskID model.LabelValue
+			var taskCpuUsage *entities.Task
+			if taskID, ok = sampleStream.Metric[s.dedicatedLabelNameTaskID]; ok {
+				if taskCpuUsage, ok = colocatedTasksCpuUsageSeconds[taskIDValueT(taskID)]; !ok {
+					// First time fetching metrics for task. Recording taskID and hostname to help consolidation.
+					taskCpuUsage = &entities.Task{
+						Metric:   sampleStream.Metric,
+						Weight:   0.0,
+						ID:       string(taskID),
+						Hostname: string(hostname),
+					}
+
+					colocatedTasksCpuUsageSeconds[taskIDValueT(taskID)] = taskCpuUsage
+				}
+
+				if _, ok = sampleStream.Metric[cpuLabelName]; ok {
+					// Calculating and recording the cpu utilization (%) of this task on this cpu.
+					// Adding it to an accumulator that will later be used as the cpu utilization of
+					// this task across all cpus.
+					taskCpuUsage.Weight += s.cpuUtil(sampleStream.Values)
+				} else {
+					// CPU usage data does not correspond to a particular cpu.
+					// TODO do not ignore this. We should instead log this.
+				}
+			}
+		} else {
+			// Either hostname not exported along with the metrics, or
+			// Task Hostname dedicated label provided is incorrect.
+			// TODO do not ignore this. We should log this instead.
+		}
+	}
+
+	// Rank colocated tasks in non-increasing order based on their total cpu utilization (%)
+	// on the entire host (all cpus).
+	rankedTasks := make(entities.RankedTasks)
+	for hostname, colocatedTasksCpuUsageSeconds := range allTasksCpuUsageSeconds {
+		for _, taskCpuUsage := range colocatedTasksCpuUsageSeconds {
+			taskCpuUsage.Weight = s.round(taskCpuUsage.Weight)
+			rankedTasks[entities.Hostname(hostname)] = append(rankedTasks[entities.Hostname(hostname)], *taskCpuUsage)
+		}
+
+		// Sorting colocated tasks.
+		sort.SliceStable(rankedTasks[entities.Hostname(hostname)], func(i, j int) bool {
+			return rankedTasks[entities.Hostname(hostname)][i].Weight >=
+				rankedTasks[entities.Hostname(hostname)][j].Weight
+		})
+	}
+
+	// Submitting ranked tasks to the receiver.
+	s.receiver.Receive(rankedTasks)
+}
+
+// cpuUtilPrecision defines the precision for cpu utilization (%) values.
+// As cpu utilization can be less than 1%, the precision is set to 2.
+const cpuUtilPrecision = 2
+
+// round the cpu utilization (%) to the above specified precision.
+func (s TaskRankCpuUtilStrategy) round(cpuUtil float64) float64 {
+	multiplier := math.Pow10(cpuUtilPrecision)
+	magnified := cpuUtil * multiplier
+	return math.Round(magnified) / multiplier
+}
+
+// cpuUtil calculates and returns the cpu utilization (%) for the task.
+// Elapsed time is calculated as the number of seconds between oldest and newest value by
+// factoring in the prometheus scrape interval.
+func (s TaskRankCpuUtilStrategy) cpuUtil(values []model.SamplePair) float64 {
```

Comment:
This will give us the average utilization in terms of cpu seconds. Is the same unit used for cpu shares too? If so, comparison between cpu utilization (of running tasks) and cpu shares (of queued tasks) will be easier.

Reply:
This gives the cpu utilization in % and not seconds.
`% = 100 * ((seconds - seconds) /  (num * seconds))`.

cpushares is a spec and therefore a constant. Just a value.
Why do we need a comparison between cpu utilization and cpu shares? We have two different strategies, one that ranks tasks based on their cpushares allocation and the other that ranks tasks based on their cpu utilization.

Reply:
Oh right, I somehow overlooked the multiplication of 100.
Ultimately, for the scheduling decision, cpu utilization gives us the right picture for scheduling decisions, right? As we discussed, we take cpu share request by a task in the queue at its face value, i.e, its utilization is assumed to be same as its cpu share. Hence for scheduling a task/job, we'll need to compare cpu utilization of the running tasks against the cpu share request for the tasks/jobs in the queue.